drivers/usb/device/usb_dc_nrf5.c

```diff
@@ -0,0 +1,2256 @@
+			continue;
+		}
+
+		if (nrf_usbd_event_check(event_map[pos])) {
+			nrf_usbd_event_clear(event_map[pos]);
+
+			evt_hdlr = isr_event_handlers[pos];
+			if (evt_hdlr) {
+				evt_hdlr(pos);
+				processed++;
+			}
+		}
+		pos++;
+	}
+
+	return processed;
+}
+
+/**
+ * @brief Interrupt service routine.
+ *
+ * This handles the USBD interrupt
+ *
+ * @param arg Argument to ISR.
+ *
+ * @return N/A
+ */
+static void nrf5_usbd_isr_handler(void *arg)
+{
+	struct nrf5_usbd_ctx *ctx = arg;
+
+	if (!ctx->enabled) {
+		return;
+	}
+
+	irq_disable(CONFIG_USBD_NRF5_IRQ);
+
+	if (nrf5_usbd_process_interrupts(ctx->enable_mask)) {
+		if (!k_work_pending(&ctx->usb_work)) {
+			k_work_submit(&ctx->usb_work);
+		}
+	}
+
+	irq_enable(CONFIG_USBD_NRF5_IRQ);
+}
+
+/**
+ * @brief Install the Interrupt service routine.
+ *
+ * This installs the interrupt service routine and enables the USBD interrupt.
+ *
+ * @return N/A
+ */
+static void nrf5_usbd_install_isr(void)
+{
+	IRQ_CONNECT(CONFIG_USBD_NRF5_IRQ,
+		    CONFIG_USBD_NRF5_IRQ_PRI,
+		    nrf5_usbd_isr_handler, &usbd_ctx, 0);
+}
+
+static void nrf5_usbd_enable_interrupts(void)
+{
+	struct nrf5_usbd_ctx *ctx = get_usbd_ctx();
+
+	ctx->enable_mask = NRF_USBD_INT_USBRESET_MASK |
+			   NRF_USBD_INT_USBEVENT_MASK |
+			   NRF_USBD_INT_ACCESSFAULT_MASK |
+			   NRF_USBD_INT_DATAEP_MASK;
+
+	nrf_usbd_int_enable(ctx->enable_mask);
+	ctx->enabled = true;
+
+	irq_enable(CONFIG_USBD_NRF5_IRQ);
+}
+
+static void nrf5_usbd_disable_interrupts(void)
+{
+	struct nrf5_usbd_ctx *ctx = get_usbd_ctx();
+
+	nrf_usbd_int_disable(~0);
+	ctx->enabled = false;
+	irq_disable(CONFIG_USBD_NRF5_IRQ);
+}
+
+void nrf5_usbd_power_event_callback(nrf_power_event_t event)
+{
+	struct nrf5_usbd_ctx *ctx = get_usbd_ctx();
+
+	switch (event) {
+	case NRF_POWER_EVENT_USBDETECTED:
+		ctx->state = USBD_ATTACHED;
+		break;
+	case NRF_POWER_EVENT_USBPWRRDY:
+		ctx->state = USBD_POWERED;
+		break;
+	case NRF_POWER_EVENT_USBREMOVED:
+		ctx->state = USBD_DETACHED;
+		break;
+	default:
+		SYS_LOG_DBG("Unknown USB event");
+		return;
+	}
+
+	SET_BIT(ctx->flags, NRF5_USB_STATE_CHANGE);
+	k_work_submit(&ctx->usb_work);
+}
+
+/**
+ * @brief Enable/Disable the HF clock
+ *
+ * Toggle the HF clock. It needs to be enabled for USBD data exchange
+ *
+ * @param on		Set true to enable the HF clock, false to disable.
+ * @param blocking	Set true to block wait till HF clock stabilizes.
+ *
+ * @return 0 on success, error number otherwise
+ */
+static int nrf5_usbd_hf_clock_enable(bool on, bool blocking)
+{
+	int ret = -ENODEV;
+	struct device *clock;
+
+	clock = device_get_binding(CONFIG_CLOCK_CONTROL_NRF5_M16SRC_DRV_NAME);
+	if (!clock) {
+		SYS_LOG_ERR("NRF5 HF Clock device not found!");
+		goto err_clk_dev;
+	}
+
+	if (on) {
+		ret = clock_control_on(clock, (void *)blocking);
+	} else {
+		ret = clock_control_off(clock, (void *)blocking);
+	}
+
+	if (ret) {
+		SYS_LOG_ERR("NRF5 HF clock %s fail: %d",
+			    on ? "start" : "stop", ret);
+		goto err_clk_start_stop;
```

Comment:
`return ret` here directly, no need for `goto`

Reply:
Ack.